### PR TITLE
Expose REST APIs

### PR DESCRIPTION
**Pull request by DimitarPetrov**
# Motivation
Rest APIs must be exposed trough controllers.

# Approach
Implement an api with different controllers each one of them contains different routes and handles http requests for this route.

### CODE REVIEW

File: web/controller.go [hunk 1]
```diff
@@ -34,3 +37,24 @@ type Endpoint struct {
...
+	Extract(*http.Request) (string, error)
+}
+
+type FuncCredentialsExtractor func(*http.Request) (string, error)
+
+func (fce FuncCredentialsExtractor) Extract(r *http.Request) (string, error) {
+	return fce(r)
+}
+
+func BasicCredentialsExtract(r *http.Request) (string, error) {
```
Comment (DimitarPetrov) [inline web/controller.go]: BasicCredentialsUsernameExtractor since you return only the username.

File: web/controller.go [hunk 2]
```diff
@@ -34,3 +37,24 @@ type Endpoint struct {
 	Method string
 	Path   string
 }
+
+type CredentialsExtractor interface {
+	Extract(*http.Request) (string, error)
+}
+
+type FuncCredentialsExtractor func(*http.Request) (string, error)
```
Comment (DimitarPetrov) [inline web/controller.go]: CredentialsExtractorFunc looks better, and dont forget the comments on exported types

File: web/controller.go [hunk 3]
```diff
@@ -34,3 +37,24 @@ type Endpoint struct {
 	Method string
 	Path   string
 }
+
+type CredentialsExtractor interface {
```
Comment (DimitarPetrov) [inline web/controller.go:32]: Comment here too

File: api/post/post_controller.go [hunk 4]
```diff
@@ -1,17 +1,17 @@
...
 	"net/http"
-	"strings"
+
+	"github.com/cc2k19/go-tin/storage"
+	"github.com/cc2k19/go-tin/web"
 )
 
 type controller struct {
 	repository *storage.Repository
+	ce         web.CredentialsExtractor
```
Comment (DimitarPetrov) [inline api/post/post_controller.go]: maybe full name since it is struct field and have bigger scope

File: api/post/post_controller.go [hunk 5]
```diff
@@ -20,23 +20,19 @@ func (c *controller) add(wr http.ResponseWriter, r *http.Request) {
...
 		log.Printf("Could not extract body: %s\n", err)
 		wr.WriteHeader(http.StatusBadRequest)
 		return
 	}
 
-	auth := r.Header.Get("Authorization")
-	decodedCredentials, err := base64.StdEncoding.DecodeString(auth[6:])
+	username, err := c.ce.Extract(r)
 	if err != nil {
 		log.Printf("Authorization decode error: %s", err)
```
Comment (DimitarPetrov) [inline api/post/post_controller.go:31]: The credentials extractor returns Authorization decore error, and you wrap it again with the same

File: api/post/post_controller.go [hunk 6]
```diff
@@ -49,16 +45,13 @@ func (c *controller) add(wr http.ResponseWriter, r *http.Request) {
 func (c *controller) get(wr http.ResponseWriter, r *http.Request) {
 	ctx := r.Context()
 
-	auth := r.Header.Get("Authorization")
-	decodedCredentials, err := base64.StdEncoding.DecodeString(auth[6:])
+	username, err := c.ce.Extract(r)
 	if err != nil {
 		log.Printf("Authorization decode error: %s", err)
```
Comment (DimitarPetrov) [inline api/post/post_controller.go:50]: Here also

File: api/post/post_controller.go [hunk 7]
```diff
@@ -1,17 +1,17 @@
...
+	"github.com/cc2k19/go-tin/storage"
+	"github.com/cc2k19/go-tin/web"
 )
 
 type controller struct {
 	repository *storage.Repository
+	ce         web.CredentialsExtractor
 }
 
 func (c *controller) add(wr http.ResponseWriter, r *http.Request) {
```
Comment (DimitarPetrov) [inline api/post/post_controller.go]: change the ResponseWriter argument name here also

File: api/user/user_controller.go [hunk 8]
```diff
@@ -1,146 +1,121 @@
...
+func (c *controller) executeRelation(r *http.Request, f func(ctx context.Context, follower string, target string) error) error {
 	defer r.Body.Close()
 
 	ctx := r.Context()
 
-	auth := r.Header.Get("Authorization")
-	decodedCredentials, err := base64.StdEncoding.DecodeString(auth[6:])
+	username, err := c.ce.Extract(r)
 	if err != nil {
 		log.Printf("Authorization decode error: %s", err)
```
Comment (DimitarPetrov) [inline api/user/user_controller.go:87]: Here also

File: api/user/user_controller.go [hunk 9]
```diff
@@ -1,146 +1,121 @@
...
 
-func (c *controller) getFollowing(wr http.ResponseWriter, r *http.Request) {
+func (c *controller) getInteraction(rw http.ResponseWriter, r *http.Request, f func(ctx context.Context, username string) (models.UserSlice, error)) {
 	ctx := r.Context()
 
-	auth := r.Header.Get("Authorization")
-	decodedCredentials, err := base64.StdEncoding.DecodeString(auth[6:])
+	username, err := c.ce.Extract(r)
 	if err != nil {
 		log.Printf("Authorization decode error: %s", err)
```
Comment (DimitarPetrov) [inline api/user/user_controller.go:109]: Here also

**Review by ataleksandrov — APPROVED** [review]

**Pull request merged**